opm/autodiff/BlackoilModelBase_impl.hpp

```diff
@@ -1121,6 +1167,88 @@ namespace detail {
+            b_perfcells_const[phase] = ADB::constant(b_perfcells[phase].value());
+        }
+        asImpl().addWellEq(wellSolutionState, well_state, mob_perfcells_const, b_perfcells_const, aliveWells,cq_s);
+        addWellControlEq(wellSolutionState, well_state, aliveWells);
+        bool converged = getWellConvergence(it);
+        while ( (!converged && (it < 15))) {
+
+            std::vector<ADB> eqs;
+            eqs.reserve(2);
+            eqs.push_back(residual_.well_flux_eq);
+            eqs.push_back(residual_.well_eq);
+            ADB total_residual = vertcatCollapseJacs(eqs);
+            const std::vector<M>& Jn = total_residual.derivative();
+            const Eigen::SparseLU< M > solver(Jn[0]);
+            const Eigen::VectorXd& dx = solver.solve(total_residual.value().matrix());
+            const int numeq = well_state.numWells()*(well_state.numPhases()+1);
+            V dx_V = V(numeq);
+            std::copy_n(dx.data(),numeq, dx_V.data());
+            updateWellState(dx_V,well_state);
+            updateWellControls(well_state);
+            //bhp and Q for the wells
+            vars0.clear();
+            variableWellStateInitials(well_state,vars0);
+            vars = ADB::variables(vars0);
+            wellSolutionState = state0;
+            variableStateExtractWellsVars(indices,vars,wellSolutionState);
+            asImpl().addWellEq(wellSolutionState, well_state, mob_perfcells_const, b_perfcells_const, aliveWells,cq_s);
+            addWellControlEq(wellSolutionState, well_state, aliveWells);
+            it++;
+            converged = getWellConvergence(it);
+        }
+        if (converged) {
+            std::cout << "well converged iter: " << it << std::endl;
+            const int nw = wells().number_of_wells;
+            {
+                // We will set the bhp primary variable to the new ones,
+                // but we do not change the derivatives here.
+                ADB::V new_bhp = Eigen::Map<ADB::V>(well_state.bhp().data(), nw);
+                // Avoiding the copy below would require a value setter method
+                // in AutoDiffBlock.
+                std::vector<ADB::M> old_derivs = state.bhp.derivative();
+                state.bhp = ADB::function(std::move(new_bhp), std::move(old_derivs));
+            }
+            {
+                // Need to reshuffle well rates, from phase running fastest
+                // to wells running fastest.
+                // The transpose() below switches the ordering.
+                const DataBlock wrates = Eigen::Map<const DataBlock>(well_state.wellRates().data(), nw, np).transpose();
+                ADB::V new_qs = Eigen::Map<const V>(wrates.data(), nw*np);
+                std::vector<ADB::M> old_derivs = state.qs.derivative();
+                state.qs = ADB::function(std::move(new_qs), std::move(old_derivs));
+            }
+            computeWellConnectionPressures(state, well_state);
```

Comment:
This call is interesting: it makes the connection pressures a bit implicit, instead of purely explicit as they have been. Was this an intentional and necessary change?